File: kb_modules/kb_typescript/postgres/kb_construct/src/ConstructStreamTable.ts
// ConstructStreamTable.ts
import { Client, QueryResult } from 'pg';
import { ConstructKB } from './ConstructKB';

export class ConstructStreamTable {
  private client: Client;
  private constructKb: ConstructKB;
  private database: string;
  private tableName: string;

  constructor(client: Client, constructKb: ConstructKB, database: string) {
    this.client = client;
    this.constructKb = constructKb;
    this.database = database;
    this.tableName = `${database}_stream`;
  }

  /** Helper to quote identifiers safely */
  private ident(id: string): string {
    return `"${id.replace(/"/g, '""')}"`;
  }

  /** Initializes schema: drops and recreates stream table and indexes */
  async setupSchema(): Promise<void> {
    const t = this.ident(this.tableName);
    // Drop table if exists
    await this.client.query(`DROP TABLE IF EXISTS ${t} CASCADE;`);
    // Enable ltree extension
    await this.client.query(`CREATE EXTENSION IF NOT EXISTS ltree;`);
    // Drop again (mirrors Python logic)
    await this.client.query(`DROP TABLE IF EXISTS ${t} CASCADE;`);

    // Create table
    await this.client.query(
      `CREATE TABLE ${t}(
         id SERIAL PRIMARY KEY,
         path LTREE,
         recorded_at TIMESTAMPTZ DEFAULT NOW(),
         valid BOOLEAN DEFAULT FALSE,
         data JSONB
       );`
    );

    // Create indexes
    const indexes = [
      [`idx_${this.tableName}_path_gist`, `USING GIST (path)`],
      [`idx_${this.tableName}_path_btree`, `(path)`],
      [`idx_${this.tableName}_recorded_at`, `(recorded_at)`],
      [`idx_${this.tableName}_recorded_at_desc`, `(recorded_at DESC)`],
      [`idx_${this.tableName}_path_recorded_at`, `(path, recorded_at)`],
    ];

    for (const [idxName, clause] of indexes) {
      const idx = this.ident(idxName);
      await this.client.query(
        `CREATE INDEX IF NOT EXISTS ${idx} ON ${t} ${clause};`
      );
    }
  }

  /**
   * Adds a stream field via the ConstructKB instance and returns a summary.
   */
  async addStreamField(
    streamKey: string,
    streamLength: number,
    description: string,
  ): Promise<{
    stream: string;
    message: string;
    properties: { stream_length: number };
    data: string;
  }> {
    if (typeof streamKey !== 'string') {
      throw new TypeError('streamKey must be a string');
    }
    if (!Number.isInteger(streamLength)) {
      throw new TypeError('streamLength must be an integer');
    }

    const properties = { stream_length: streamLength };
    // Delegate to ConstructKB
    await this.constructKb.addInfoNode(
      'KB_STREAM_FIELD',
      streamKey,
      properties,
      {},
      description,
    );

    return {
      stream: 'success',
      message: `stream field '${streamKey}' added successfully`,
      properties,
      data: description,
    };
  }

  /**
   * Removes entries whose paths are in invalidStreamPaths, in chunks.
   */
  private async removeInvalidStreamFields(
    invalidStreamPaths: string[],
    chunkSize = 500,
  ): Promise<void> {
    if (invalidStreamPaths.length === 0) return;

    const t = this.ident(this.tableName);
    for (let i = 0; i < invalidStreamPaths.length; i += chunkSize) {
      const chunk = invalidStreamPaths.slice(i, i + chunkSize);
      const placeholders = chunk.map((_, idx) => `$${idx + 1}`).join(', ');
      const sql = `DELETE FROM ${t} WHERE path IN (${placeholders});`;
      await this.client.query(sql, chunk);
    }
  }

  /**
   * Ensures each path has exactly the target number of rows: deletes oldest or inserts new.
   */
  private async manageStreamTable(
    paths: string[],
    lengths: number[],
  ): Promise<void> {
    const t = this.ident(this.tableName);

    for (let i = 0; i < paths.length; i++) {
      const path = paths[i];
      const target = lengths[i];

      // count current rows
      const countRes: QueryResult = await this.client.query(
        `SELECT COUNT(*) AS cnt FROM ${t} WHERE path = $1;`,
        [path],
      );
      const current = parseInt((countRes.rows[0].cnt as unknown as string), 10);
      const diff = target - current;

      if (diff < 0) {
        // delete oldest
        const delSql = `
          DELETE FROM ${t}
          WHERE path = $1
            AND recorded_at IN (
              SELECT recorded_at FROM ${t}
              WHERE path = $1
              ORDER BY recorded_at ASC
              LIMIT $2
            );
        `;
        await this.client.query(delSql, [path, Math.abs(diff)]);
      } else if (diff > 0) {
        // insert placeholders
        const insSql = `INSERT INTO ${t} (path, recorded_at, data, valid)
                        VALUES ($1, NOW(), $2, FALSE);`;
        for (let j = 0; j < diff; j++) {
          await this.client.query(insSql, [path, {}]);
        }
      }
    }
  }

  /**
   * Synchronize stream table with KB entries: remove invalid, add missing.
   */
  async checkInstallation(): Promise<void> {
    const t = this.ident(this.tableName);

    // fetch existing stream paths
    const spRes = await this.client.query(
      `SELECT DISTINCT path::text AS path FROM ${t};`
    );
    const existingPaths = spRes.rows.map(r => r.path as string);

    // fetch KB-defined stream fields
    const kbTable = this.ident(this.database);
    const kbRes = await this.client.query(
      `SELECT path, properties FROM ${kbTable}
       WHERE label = 'KB_STREAM_FIELD';`
    );
    const definedPaths = kbRes.rows.map(r => r.path as string);
    const definedLengths = kbRes.rows.map(
      r => (r.properties.stream_length as number)
    );

    // compute diffs
    const invalid = existingPaths.filter(p => !definedPaths.includes(p));
    const missing = definedPaths.filter(p => !existingPaths.includes(p));

    await this.removeInvalidStreamFields(invalid);
    await this.manageStreamTable(definedPaths, definedLengths);
  }
}

File: kb_modules/kb_rustc/kb_create/kb_create_module/src/ConstructStreamTable.rs
use std::collections::HashMap;
use postgres::{Client, Error, NoTls, types::Json, Row};
use serde_json::Value;
use crate::ConstructKb;

/// Manages a stream table tied to a ConstructKb, setting up schema and
/// synchronizing stream entries based on KB-defined stream fields.
pub struct ConstructStreamTable {
    client: Client,
    construct_kb: ConstructKb,
    database: String,
    table_name: String,
}

impl ConstructStreamTable {
    /// Create a new stream-table manager, dropping/re-creating schema.
    pub fn new(
        mut client: Client,
        construct_kb: ConstructKb,
        database: &str,
    ) -> Result<Self, Error> {
        let table_name = format!("{}_stream", database);
        let mut inst = ConstructStreamTable { client, construct_kb, database: database.into(), table_name };
        inst.setup_schema()?;
        Ok(inst)
    }

    fn setup_schema(&mut self) -> Result<(), Error> {
        // Drop existing
        let drop = format!("DROP TABLE IF EXISTS \"{}\" CASCADE;", self.table_name);
        self.client.batch_execute(&drop)?;

        // Ensure ltree
        self.client.batch_execute("CREATE EXTENSION IF NOT EXISTS ltree;")?;

        // Drop again (optional mirror of Python)
        self.client.batch_execute(&drop)?;

        // Create table
        let create = format!(r#"
            CREATE TABLE "{}" (
                id SERIAL PRIMARY KEY,
                path LTREE,
                recorded_at TIMESTAMPTZ DEFAULT NOW(),
                valid BOOLEAN DEFAULT FALSE,
                data JSONB
            );
        "#, self.table_name);
        self.client.batch_execute(&create)?;

        // Indexes
        let idxs = vec![
            format!("CREATE INDEX IF NOT EXISTS idx_{}_path_gist ON \"{}\" USING GIST(path);", self.table_name, self.table_name),
            format!("CREATE INDEX IF NOT EXISTS idx_{}_path_btree ON \"{}\"(path);", self.table_name, self.table_name),
            format!("CREATE INDEX IF NOT EXISTS idx_{}_recorded_at ON \"{}\"(recorded_at);", self.table_name, self.table_name),
            format!("CREATE INDEX IF NOT EXISTS idx_{}_recorded_at_desc ON \"{}\"(recorded_at DESC);", self.table_name, self.table_name),
            format!("CREATE INDEX IF NOT EXISTS idx_{}_path_recorded_at ON \"{}\"(path, recorded_at);", self.table_name, self.table_name),
        ];
        for sql in idxs { self.client.batch_execute(&sql)?; }
        Ok(())
    }

    /// Defines a new stream field in KB and returns a summary JSON.
    pub fn add_stream_field(
        &mut self,
        stream_key: &str,
        stream_length: i32,
        description: &str,
    ) -> Result<Value, Error> {
        // Add KB info node
        let mut props = HashMap::new();
        props.insert("stream_length".to_string(), Value::Number(stream_length.into()));
        // NOTE: empty data object
        let data = Value::Object(Default::default());
        self.construct_kb.add_info_node(
            "KB_STREAM_FIELD",
            stream_key,
            &mut props,
            &data,
            Some(description),
        )?;

        // Build result JSON
        Ok(json!({
            "stream": "success",
            "message": format!("stream field '{}' added successfully", stream_key),
            "properties": props,
            "data": description,
        }))
    }

    /// Remove all entries whose path matches any in `invalid_paths`, in chunks.
    pub fn remove_invalid_stream_fields(
        &mut self,
        invalid_paths: &[String],
        chunk_size: usize,
    ) -> Result<(), Error> {
        if invalid_paths.is_empty() { return Ok(()); }

        for chunk in invalid_paths.chunks(chunk_size) {
            // Build placeholders $1,$2,...
            let placeholders: Vec<String> = (1..=chunk.len()).map(|i| format!("${}", i)).collect();
            let in_clause = placeholders.join(",");
            let stmt = format!(
                "DELETE FROM \"{}\" WHERE path IN ({});",
                self.table_name,
                in_clause
            );
            // Convert chunk to &[&(dyn ToSql)]
            let params: Vec<&(dyn postgres::types::ToSql + Sync)> =
                chunk.iter().map(|s| s as &(dyn postgres::types::ToSql + Sync)).collect();
            self.client.execute(stmt.as_str(), &params)?;
        }
        Ok(())
    }

    /// Ensures each path has exactly the specified number of records.
    pub fn manage_stream_table(
        &mut self,
        specified_paths: &[String],
        specified_lengths: &[i32],
    ) -> Result<(), Error> {
        for (path, &target) in specified_paths.iter().zip(specified_lengths) {
            // Count existing
            let count_stmt = format!("SELECT COUNT(*) FROM \"{}\" WHERE path = $1;", self.table_name);
            let row = self.client.query_one(&count_stmt, &[&path])?;
            let current: i64 = row.get(0);
            let diff = target as i64 - current;

            if diff < 0 {
                // delete oldest |diff|
                let del = format!(r#"
                    DELETE FROM "{}"
                    WHERE path = $1 AND recorded_at IN (
                        SELECT recorded_at FROM "{}"
                        WHERE path = $1 ORDER BY recorded_at ASC LIMIT $2
                    );
                "#, self.table_name, self.table_name);
                self.client.execute(&del, &[&path, &(-diff as i64)])?;
            } else if diff > 0 {
                // insert diff new null entries
                let ins = format!(r#"
                    INSERT INTO "{}" (path, recorded_at, data, valid)
                    VALUES ($1, NOW(), $2, FALSE);
                "#, self.table_name);
                for _ in 0..diff {
                    // JSONB null => Value::Null
                    self.client.execute(&ins, &[&path, &Json(&Value::Null)])?;
                }
            }
        }
        Ok(())
    }

    /// Synchronize stream table with KB definitions.
    pub fn check_installation(&mut self) -> Result<(), Error> {
        // 1) fetch distinct stream paths
        let p = format!("SELECT DISTINCT path::text FROM \"{}\";", self.table_name);
        let rows: Vec<Row> = self.client.query(&p, &[])?;
        let unique_paths: Vec<String> = rows.iter().map(|r| r.get(0)).collect();

        // 2) fetch KB-defined fields
        let kq = format!(
            "SELECT path, properties->>'stream_length' as sl FROM \"{}\" WHERE label='KB_STREAM_FIELD';",
            self.database
        );
        let krows: Vec<Row> = self.client.query(&kq, &[])?;
        let mut specified_paths = Vec::new();
        let mut specified_lengths = Vec::new();
        for r in krows {
            let p: String = r.get(0);
            let sl: i32 = r.get::<_, String>(1).parse().unwrap_or(0);
            specified_paths.push(p);
            specified_lengths.push(sl);
        }

        // 3) diff
        let invalid: Vec<String> = unique_paths.iter().filter(|p| !specified_paths.contains(p)).cloned().collect();

        self.remove_invalid_stream_fields(&invalid, 500)?;
        self.manage_stream_table(&specified_paths, &specified_lengths)?;
        Ok(())
    }
}
